### PR TITLE
feat(x/data): query iri by hash cli and rest endpoint

### PR DESCRIPTION
**Pull request by ryanchristo**
## Description

Closes: #1087

This pull request adds a CLI command to query IRI by content hash using a json file as well as the ability to query IRI by raw and graph content hash properties using a gRPC REST endpoint.

### CLI Command

```
./build/regen q data iri content.json
```

```
{
  "graph": {
    "hash": "YWJjZGVmZ2hpamtsbW5vcHFyc3R1dnd4eXoxMjM0NTY=",
    "digest_algorithm": "DIGEST_ALGORITHM_BLAKE2B_256",
    "canonicalization_algorithm": "GRAPH_CANONICALIZATION_ALGORITHM_URDNA2015",
    "merkle_tree": "GRAPH_MERKLE_TREE_NONE_UNSPECIFIED"
  }
}
```

### REST Endpoints

```
curl "localhost:1317/regen/data/v1/iri-by-raw?hash=YWJjZGVmZ2hpamtsbW5vcHFyc3R1dnd4eXoxMjM0NTY=&digest_algorithm=1"
```

```
curl "localhost:1317/regen/data/v1/iri-by-raw?hash=YWJjZGVmZ2hpamtsbW5vcHFyc3R1dnd4eXoxMjM0NTY=&digest_algorithm=DIGEST_ALGORITHM_BLAKE2B_256"
```

```
curl "localhost:1317/regen/data/v1/iri-by-graph?hash=YWJjZGVmZ2hpamtsbW5vcHFyc3R1dnd4eXoxMjM0NTY=&digest_algorithm=1&canonicalization_algorithm=1"
```

```
curl "localhost:1317/regen/data/v1/iri-by-graph?hash=YWJjZGVmZ2hpamtsbW5vcHFyc3R1dnd4eXoxMjM0NTY=&digest_algorithm=DIGEST_ALGORITHM_BLAKE2B_256&canonicalization_algorithm=GRAPH_CANONICALIZATION_ALGORITHM_URDNA2015"
```

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

**Comment by ryanchristo** [comment]
cc @paul121 if you want to take a look as well.

**Review by aleem1314 — APPROVED** [review]
lgtm

File: x/data/client/testsuite/grpc.go [hunk 1]
```diff
@@ -159,6 +161,249 @@ func (s *IntegrationTestSuite) TestQueryHashByIRI() {
...
+		s.Run(tc.name, func() {
+			bz, err := rest.GetRequest(tc.url)
+			require.NoError(err)
+
+			var res data.QueryIRIByRawHashResponse
+			err = val.ClientCtx.Codec.UnmarshalJSON(bz, &res)
+
+			if tc.expErr {
+				require.Error(err)
+				require.Contains(string(bz), tc.errMsg)
```
Comment (technicallyty) [inline x/data/client/testsuite/grpc.go:269]: should be able to use `ErrorContains` here

File: x/data/client/testsuite/grpc.go [hunk 2]
```diff
@@ -159,6 +161,249 @@ func (s *IntegrationTestSuite) TestQueryHashByIRI() {
...
+		s.Run(tc.name, func() {
+			bz, err := rest.GetRequest(tc.url)
+			require.NoError(err)
+
+			var res data.QueryIRIByGraphHashResponse
+			err = val.ClientCtx.Codec.UnmarshalJSON(bz, &res)
+
+			if tc.expErr {
+				require.Error(err)
+				require.Contains(string(bz), tc.errMsg)
```
Comment (technicallyty) [inline x/data/client/testsuite/grpc.go:398]: ditto

File: x/data/client/testsuite/grpc.go [hunk 3]
```diff
@@ -329,3 +574,10 @@ func (s *IntegrationTestSuite) TestQueryResolvers() {
 		})
 	}
 }
+
+func encodeBase64Bytes(decodedHash []byte) string {
+	// encode base64 bytes to base64 string
+	encodedHash := base64.StdEncoding.EncodeToString(decodedHash)
+	// replace all instances of "+" with "%2b"
+	return strings.Replace(encodedHash, "+", "%2b", -1)
```
Comment (technicallyty) [inline x/data/client/testsuite/grpc.go:582]: just curious, why do we do the replace?

**Review by technicallyty — APPROVED** [review]
pre-approving, few nits 👍🏻

File: content.json [hunk 4]
```diff
@@ -0,0 +1,8 @@
+    {
+      "graph": {
+        "hash": "YWJjZGVmZ2hpamtsbW5vcHFyc3R1dnd4eXoxMjM0NTY=",
+        "digest_algorithm": "DIGEST_ALGORITHM_BLAKE2B_256",
```
[the remark this answers is not in the document]
Comment (ryanchristo) [reply content.json]: https://github.com/regen-network/regen-ledger/pull/1107/commits/98ceb6f7f0779f425b539976a815648d035a8e00

File: x/data/client/testsuite/grpc.go (hunk 1 above)
Comment (ryanchristo) [reply x/data/client/testsuite/grpc.go:269]: `bz` is the response from `GetRequest` so we are checking if the error message is within the get response. The `err` above is from `UnmarshalJSON` so we need to check if the response contains the error message here.

File: x/data/client/testsuite/grpc.go (hunk 2 above)
Comment (ryanchristo) [reply x/data/client/testsuite/grpc.go:398]: Same as above. We need to check if the response contains the error message.

File: x/data/client/testsuite/grpc.go (hunk 3 above)
Comment (ryanchristo) [reply x/data/client/testsuite/grpc.go:582]: For REST endpoints, we cannot use `+` in the param value so it needs to be encoded.

**Pull request merged**